File: app/src/main/java/com/kinwatt/powermeter/ui/fragments/DeviceRecyclerViewAdapter.java
package com.kinwatt.powermeter.ui.fragments;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.kinwatt.powermeter.R;
import com.kinwatt.powermeter.data.SensorData;
import com.kinwatt.powermeter.ui.fragments.DeviceFragment.OnListFragmentInteractionListener;

import java.util.List;

public class DeviceRecyclerViewAdapter extends RecyclerView.Adapter<DeviceRecyclerViewAdapter.DeviceViewHolder> {

    private final List<SensorData> items;
    private final OnListFragmentInteractionListener listener;

    public DeviceRecyclerViewAdapter(List<SensorData> items, OnListFragmentInteractionListener listener) {
        this.items = items;
        this.listener = listener;
    }

    @Override
    public DeviceViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new DeviceViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.fragment_device, parent, false));
    }

    @Override
    public void onBindViewHolder(final DeviceViewHolder holder, int position) {
        holder.setItem(items.get(position));
        holder.setOnClickListener(v -> {
            if (listener != null) {
                listener.onListFragmentInteraction(holder.getItem());
            }
        });
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public class DeviceViewHolder extends RecyclerView.ViewHolder {
        private final View view;
        private final TextView deviceName;
        private SensorData item;

        public DeviceViewHolder(View view) {
            super(view);
            this.view = view;
            this.deviceName = view.findViewById(R.id.device_name);
        }

        public SensorData getItem() {
            return item;
        }
        public void setItem(SensorData item) {
            this.item = item;
            this.deviceName.setText(item.getName());
        }

        public void setOnClickListener(View.OnClickListener listener) {
            view.setOnClickListener(listener);
        }

        @Override
        public String toString() {
            return super.toString() + " '" + item.getName() + "'";
        }
    }
}

File: app/src/main/java/com/kinwatt/powermeter/ui/fragments/DeviceRecyclerViewAdapter.kt
package com.kinwatt.powermeter.ui.fragments

import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.kinwatt.powermeter.R
import com.kinwatt.powermeter.data.SensorData
import com.kinwatt.powermeter.ui.fragments.DeviceFragment.OnListFragmentInteractionListener

class DeviceRecyclerViewAdapter(
    private val items: List<SensorData>,
    private val listener: OnListFragmentInteractionListener?
) : RecyclerView.Adapter<DeviceRecyclerViewAdapter.DeviceViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DeviceViewHolder {
        return DeviceViewHolder(
            LayoutInflater.from(parent.context)
                .inflate(R.layout.fragment_device, parent, false)
        )
    }

    override fun onBindViewHolder(holder: DeviceViewHolder, position: Int) {
        holder.item = items[position]
        holder.setOnClickListener( View.OnClickListener {
            listener?.onListFragmentInteraction(holder.item!!)
        } )
    }

    override fun getItemCount(): Int {
        return items.size
    }

    inner class DeviceViewHolder(private val view: View) : RecyclerView.ViewHolder(view) {
        private val deviceName: TextView = view.findViewById(R.id.device_name)

        var item: SensorData = SensorData()
            set(item) {
                field = item
                this.deviceName.text = item.name
            }

        fun setOnClickListener(listener: View.OnClickListener) {
            view.setOnClickListener(listener)
        }

        override fun toString() = "${super.toString()} '${item.name}'"
    }
}
